### ISSUE TITLE
Regression: Unable to download large files via intercept after telepresence 2.3

### ISSUE THREAD

**Issue by wojciechka** [telepresenceio/telepresence#1810]
Hello,

I have been using telepresence v2.2.2 with a GKE cluster and an nginx ingress for a few weeks, using telepresence to develop frontend against backend already in the cluster.

The frontend is using react and TypeScript, using yarn and `yarn.lock` file to lock all the dependencies explicitly. We're using webpack dev server with hot reload for development and that is how telepresence was used - original deployment of the prod version of the app gets intercepted and all requests for UI are routed to the webpack dev server on developer's computer.

The application in GKE is behind nginx-ingress (version 0.32.0) and ingress is exposing the application over https.

Things have worked fine for a while, including the hot reload feature. Recently I have updated to telepresence v2.3.2 and things stopped working. Specifically the UI never finishes loading. I can see some of webpack messages, but nothing finishes loading.

After a lot of trial and error with things that could possibly go wrong, I have ruled out:
* updates to the backend code - by rolling back to version that I am confident worked with telepresence and ensuring all backend components use a specific container image and settings (the project has a separate config repository for Kubernetes so it is easy to roll back to a specific version)
* updates to the frontend code - I have rolled back to commit I know to work with telepresence ; since all the npm packages are `yarn.lock`-ed, it should be a reproducible setup
* browser issues - I have run an older version of Chrome for a while to ensure the problem is not related to an update in the browser
* generic networking issues - I have replaced webpack dev server with `nginx` running on my machine and hosting a static directory, I was able to download small and large files without issues

After ruling out all the issues that could be caused by changes on my end, I have consistently tried all versions of telepresence and came up with `v2.2.2` working as expected and `v2.3.0` not working. I have also tried `v2.3.1` and `v2.3.2` and all of them fail to work and with `telepresence intercept` enabled.

I suspect this has to be some edge case related to many features webpack dev server is using - such as websockets it uses to inform the browser of updates in the remote code and need to reload.

Could anyone help determine what could be the issue? Is this a known problem with v2.3?

Are there some settings, additional configuration or changes to `traffic-manager` deployment I could make to get additional logs / information?

**Comment by LukeShu** [comment]
That's weird that it doesn't work but doesn't appear to be generic networking issues (since generic networking issues are what I would suspect for regressions in 2.3.0 because of all the changes in that area in that version).

Grasping at straws for "websockets": I don't suppose `telepresence status` indicates that you're logged in?  It's a known issue that the smart agent does not work with websockets (that should be addressed soon-ish).  But I'm not sure that makes sense if it was working with 2.2.2.

/cc @thallgren

**Comment by wojciechka** [comment]
> That's weird that it doesn't work but doesn't appear to be generic networking issues (since generic networking issues are what I would suspect for regressions in 2.3.0 because of all the changes in that area in that version).

Then it seems I ran into some regression here. Happy to help more with debugging and/or troubleshooting with 2.3 versions.

I'm just not very familiar with telepresence 2.x, so any tips on getting more logs or debugging this would be useful.

I've been using earlier versions of telepresence years ago and just adopting the new version nowadays.

> Grasping at straws for "websockets": I don't suppose `telepresence status` indicates that you're logged in? It's a known issue that the smart agent does not work with websockets (that should be addressed soon-ish). But I'm not sure that makes sense if it was working with 2.2.2.

If you logged in you mean Ambassador Cloud, then no, I have not signed up and just using the tool as is so my `telepresence status` reports I am `Logged out`.

Would this change anything? Either in behavior or in log gathering?

**Comment by thallgren** [comment]
You'll find the logs under `"/.cache/telepresence/logs` on a Linux box and under `~/Library/Logs/telepresence` on a macOS box. For more info regarding logging, please take a look at [Log Levels](https://www.telepresence.io/docs/latest/reference/config/#log-levels). Changing to `rootDaemon: debug` can be very helpful for cases like this.

We did have issues with GKE clusters in 2.3.0 and 2.3.1, but they should be fixed in 2.3.2.

**Comment by wojciechka** [comment]
> You'll find the logs under `"/.cache/telepresence/logs` on a Linux box and under `~/Library/Logs/telepresence` on a macOS box. For more info regarding logging, please take a look at [Log Levels](https://www.telepresence.io/docs/latest/reference/config/#log-levels). Changing to `rootDaemon: debug` can be very helpful for cases like this.
> 
> We did have issues with GKE clusters in 2.3.0 and 2.3.1, but they should be fixed in 2.3.2.

Hi.

I have retried the whole procedure on a fresh GKE cluster and I have tried to gather more information on the setup itself:

- GKE version 1.19.9-gke.1900 ; node types N2 (4 CPU, 16 GB RAM), us-east1-c region
- nginx ingress controller version 0.32.0
- UI component versions can be found here: https://github.com/influxdata/ui/blob/master/package.json and https://github.com/influxdata/ui/blob/master/yarn.lock

I have tried telepresence versions 2.2.2 and 2.3.3 and I could confirm the same pattern - when doing `telepresence intercept` with version 2.2.2 the UI works fine whereas version 2.3.3 just freezes.

I have configured log level for `rootDaemon` to `debug` and gathered logs from 2.2.2 and 2.3.3. I noticed the logs are very different, from what I understand this is a difference in behavior between 2.2.2 and 2.3.0.

I did not see anything that would be clearly an error, but I have found a few lines that indicate an error, although I think they are not related:

```
2021/07/08 10:43:43.6764 debug   daemon/server-router/MGR stream : MGR read loop starting (from pkg/client/daemon/tunrouter.go:342)
2021/07/08 10:43:43.6780 debug   daemon/server-router/TUN writer : -> TUN icmp fe80::506c:7b41:7778:97f2 -> ff02::16, type destination unreachable, code 2 (from pkg/client/daemon/tunrouter.go:312)
2021/07/08 10:43:43.6781 debug   daemon/server-router/TUN writer : -> TUN icmp fe80::506c:7b41:7778:97f2 -> ff02::16, type destination unreachable, code 2 (from pkg/client/daemon/tunrouter.go:312)
2021/07/08 10:43:43.6782 debug   daemon/server-router/TUN writer : -> TUN icmp fe80::506c:7b41:7778:97f2 -> ff02::16, type destination unreachable, code 2 (from pkg/client/daemon/tunrouter.go:312)
2021/07/08 10:43:43.6782 debug   daemon/server-router/TUN writer : -> TUN icmp fe80::506c:7b41:7778:97f2 -> ff02::16, type destination unreachable, code 2 (from pkg/client/daemon/tunrouter.go:312)
```

I will try to prepare a way to reproduce the problem outside of our environment and share it.

**Comment by wojciechka** [comment]
I have found a way to replicate most of the behavior, except that the happy path ends with an empty HTML page and a JS error - but it is distinguishable from when the problem with intercept / websockets happens.

Tested on Linux x64 with Debian 10 host (kernel 4.15.0-144-generic).

## Setting up webpack dev server locally (I was using `node` version `v12.22.1`, yarn version `v1.22.10`).

- (prerequisite) install node.js and install `yarn` if not yet installed
- clone https://github.com/influxdata/ui and cd into `ui` directory (I have a checkout of commit `e9316812d0178518437acd1a87dc2d35481dc579` but it should not depend on specific commit being used)
- run `yarn` to install all dependencies
- run `yarn start` to run the webpack dev server

## Deploying ingress and dummy workload into GKE cluster

These steps assume a cluster is already created. I was testing it with GKE, but I am not sure if a GKE cluster specifcally is needed.

This includes 3 files:
- `step1.yml` - most of ingress-nginx setup, except for webhook
- `step2.yml` - webhook for nginx
- `step3.yml` - payload for Ingress object and a dummy UI application (using `bitnami/nginx` container image that's listening on port 9999)

- edit the `step3.yml` file and change `publicly-available-hostname.domain` to a valid domain you can point to the ingress-nginx's LoadBalancer IP (or at the of this, edit local `/etc/hosts` and add `(loadbalancer ip) publicly-available-hostname.domain` to make it accessible locally)
- create namespaces `kubectl create ns ingress-nginx ; kubectl create ns ui-test` (the yaml files include namespaces, but also objects in those namespaces so in some edge cases it could lead to errors otherwise)
- run `kubectl apply -f step1.yml ; sleep 20 ; kubectl apply -f step2.yml ; sleep 30 ; kubectl apply -f step3.yml` (`sleep` s are added to avoid edge cases where nginx is not ready and its webhook will reject creation of `Ingress` object)
- run `kubectl get svc -n ingress-nginx ingress-nginx-controller` to get the external IP address of ingress, either edit DNS for your domain or `/etc/hosts` and point the domain from first step in this section to this IP address
- validate that going to https://(domain name) works and complains about invalid TLS certificate (where `(domain name)` refers to either default `publicly-available-hostname.domain` host from Ingress in `step3.yml` or the domain that was in `step3.yml`)

## Testing telepresence

- run `telepresence connect`
- run `telepresence intercept --namespace ui-test --port 8080:9999 influxdb-ui` to intercept local traffic (at this point the UI from "Setting up webpack dev server locally" above should already be running)
- open dev tools in your browser and go to https://(domain name)

If telepresence is working (i.e. v2.2.2), you should see an error related to `index.tsx` (such as `Error importing 'index.tsx': TypeError: Object(...) is not a function`) and loading should finish after a few seconds with a blank document. This is probably because the UI application is not fully configured, but it is not preventing troubleshooting or reproduction of the problem itself.

If there is an error (i.e. with v2.3.3), the page will continue to load the page for very long time and never show the error message ; after a while the console should show an `[WDS] Disconnected!` error, indicating the websocket was disconnected. The lack of `TypeError` above indicates the page was not loaded correctly. In some cases I am also getting `Error importing `index.tsx`: TypeError: Could not download wasm module` or other errors.

[step1.yml.txt](https://github.com/telepresenceio/telepresence/files/6783486/step1.yml.txt)
[step2.yml.txt](https://github.com/telepresenceio/telepresence/files/6783487/step2.yml.txt)
[step3.yml.txt](https://github.com/telepresenceio/telepresence/files/6783488/step3.yml.txt)

**Comment by thallgren** [comment]
Would it be possible for you to provide the connector.log, daemon.log, and the logs from the traffic-manager pod after a run wit the above reproducer (I'd rather look at them first before setting everything up)?

**Comment by wojciechka** [comment]
> Would it be possible for you to provide the connector.log, daemon.log, and the logs from the traffic-manager pod after a run wit the above reproducer (I'd rather look at them first before setting everything up)?

Sure. here they are, both from 2.2.2 and 2.3.3:

[v2.2.2-connector.log.txt](https://github.com/telepresenceio/telepresence/files/6784688/v2.2.2-connector.log.txt)
[v2.2.2-daemon.log.txt](https://github.com/telepresenceio/telepresence/files/6784689/v2.2.2-daemon.log.txt)
[v2.2.2-traffic-manager-logs.txt](https://github.com/telepresenceio/telepresence/files/6784690/v2.2.2-traffic-manager-logs.txt)

[v2.3.3-connector.log.txt](https://github.com/telepresenceio/telepresence/files/6784691/v2.3.3-connector.log.txt)
[v2.3.3-daemon.log.txt](https://github.com/telepresenceio/telepresence/files/6784692/v2.3.3-daemon.log.txt)
[v2.3.3-traffic-manager-logs.txt](https://github.com/telepresenceio/telepresence/files/6784693/v2.3.3-traffic-manager-logs.txt)

**Comment by thallgren** [comment]
Thanks @wojciechka . This is very helpful.

**Comment by wojciechka** [comment]
Hi, is there anything else I could do to help with the issue? I've seen a few releases, I could give them a try, but I would need to set up a new cluster, so I wanted to first check if changes between 2.3.3 and 2.3.6 could include anything related to this issue?

**Comment by thallgren** [comment]
We have had some problems with DNS that were fixed in the latest 2.3.6 release but I doubt it is relevant to the problem you have here. We have some work planned to make websockets work OK though, so perhaps wait for that to arrive?

**Comment by wojciechka** [comment]
> We have had some problems with DNS that were fixed in the latest 2.3.6 release but I doubt it is relevant to the problem you have here. We have some work planned to make websockets work OK though, so perhaps wait for that to arrive?

Thank for for the prompt response. Yes, I am happy to keep on using 2.2.2 and wait for further changes, I was simply making sure I did not miss on anything around this.

If you have a PR or any work that should improve it, I'm more than happy to help test if it solves my specific problem.

**Comment by alberthuang24** [comment]
I have the same problem when I close the websocket.

**Comment by alberthuang24** [comment]
This happens when the request return value is too large.

![image](https://user-images.githubusercontent.com/20190812/126676121-62239375-ccec-40e0-abe3-d00abdea0d54.png)

**Comment by alberthuang24** [comment]
Sorry, my problem seems to be fixed in the latest version. https://github.com/telepresenceio/telepresence/pull/1788

**Comment by wojciechka** [comment]
Just confirmed that this is resolved and telepresence 2.4.0 works with the webpack dev server setup that we have.

However, after some more tests it seems that there are still cases where the browser freezes.

I have managed to reproduce it when forcing reload of the page 2-3 times. The download gets stuck on one or more of the assets - so far it has been one of the larger JavaScript files.

It is definitely better than with version 2.3, but unfortunately it's still not working as well as 2.2.2.

I will perform more tests and see if I can find a pattern or understand when things work vs not work.

**Comment by josecv** [comment]
Okay so after a long investigation, here's what I think is happening; I'll start with some detailed steps on how I reproduce the problem, then my analysis of it.

## Repro

1. First you need some kind of local server that will serve large files; I did this in the simplest way possible:
    ```bash
    mkdir web
    dd if=/dev/random count=4k bs=5120 | base64 > web/file
    docker run --rm -it -e FOLDER=/web -v $(pwd)/web:/web -p 8080:8080 halverneus/static-file-server
    ```
1. Now we need two services -- one that will serve the large files and one that will consume them. I did this by `kubectl apply -f`ing something like:
    ```yaml
    ---
    apiVersion: v1
    kind: Service
    metadata:
      name: echo-easy
    spec:
      type: ClusterIP
      selector:
        service: echo-easy
      ports:
        - name: proxied
          port: 80
          targetPort: http
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: echo-easy
      labels:
        service: echo-easy
    spec:
      replicas: 1
      selector:
        matchLabels:
          service: echo-easy
      template:
        metadata:
          labels:
            service: echo-easy
        spec:
          containers:
            - name: echo-easy
              image: jmalloc/echo-server
              ports:
                - containerPort: 8080
                  name: http
              resources:
                limits:
                  cpu: 50m
                  memory: 128Mi
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      labels:
        app: wrk
      name: wrk
    spec:
      replicas: 1
      selector:
        matchLabels:
          app: wrk
      template:
        metadata:
          labels:
            app: wrk
        spec:
          containers:
          - command:
            - sleep
            - "6000000"
            image: skandyla/wrk
            imagePullPolicy: Always
            name: wrk
            resources: {}
          restartPolicy: Always
    ```
1. Now just intercept the service:
    ```bash
    telepresence connect
    telepresence intercept echo-easy --port 8080
    ```
1. To generate some in-cluster traffic to the intercepted service:
    ```bash
    kubectl get pod -lapp=wrk # Note the pod name, then...
    kubectl exec -ti ${WRK_POD_NAME} -- wrk -c 2 -t 2 -d 2m http://echo-easy.default/file
    # wrk is an http benchmarker, these options will launch 2 concurrent connections over 2 threads for 2 minutes.
    ```
1. Eventually wrk's requests will start timing out; by the time `wrk` is done running, even `curl http://echo-easy.default/file` will hang forever until telepresence is restarted.

## Investigation

### Why do responses stop working?

Because the agent gets itself into a hang. This in turn happens because the HTTP forwarder gets itself into a `select{}` lock.
Look at this goroutine:

```
goroutine 57 [select, 3 minutes]:
github.com/telepresenceio/telepresence/v2/pkg/connpool.(*dialer).HandleMessage(0xc0002d7560, {0x1c9fc78, 0xc0000a64c0}, {0x1c8e7f0, 0xc000306020})
	github.com/telepresenceio/telepresence/v2/pkg/connpool/dialer.go:159 +0xd6
github.com/telepresenceio/telepresence/v2/pkg/forwarder.(*Forwarder).startManagerTunnel.func2()
	github.com/telepresenceio/telepresence/v2/pkg/forwarder/forwarder.go:285 +0x13e
created by github.com/telepresenceio/telepresence/v2/pkg/forwarder.(*Forwarder).startManagerTunnel
	github.com/telepresenceio/telepresence/v2/pkg/forwarder/forwarder.go:261 +0x3ef
```

That goroutine is responsible for handling messages from the traffic manager; if it gets stuck handling a message (because there's nobody listening for the message), it'll lock the entire agent and prevent any new requests from being fulfilled.

### Why is HandleMessage writing to a channel with no receivers?

Because it's the wrong type of handler.
If I make a request to the agent, say from a pod on the cluster (such as an ingress controller), the agent will first [accept the request](https://github.com/telepresenceio/telepresence/blob/release/v2/pkg/forwarder/forwarder.go#L78), thus creating a `net.TCPConn` for it, then it will [allocate a new handler using HandlerFromConn](https://github.com/telepresenceio/telepresence/blob/release/v2/pkg/forwarder/forwarder.go#L301-L304).
When a message comes in from the traffic-agent (i.e. from the daemon on the user's machine), it will be handled by that handler, and will eventually be written out to the underlying connection.

However, if the remote end (the pod on the cluster) hangs up, the connection will die and the handler will die with it.
In that case, when a message comes in from the traffic-manager, a [new handler will be allocated](https://github.com/telepresenceio/telepresence/blob/release/v2/pkg/forwarder/forwarder.go#L278-L285) but it will never have anywhere to write to.

### Why is the traffic-manager sending a message when the connection is closed?

The simple answer is that it has no awareness that the remote end has hung up.
The traffic-manager's AgentTunnel is [pretty straightforward](https://github.com/telepresenceio/telepresence/blob/release/v2/cmd/traffic/cmd/manager/internal/state/state.go#L716-L753) -- it grabs messages from the agent and dumps them into the client tunnel for the daemon to deal with.

The slightly more complicated answer is that there are still messages to send, because they have accumulated on the client tunnel.
Look at these logs (lightly edited for readability):
```
# daemon.log
2021/09/03 12:34:27.8131 debug   daemon/server-router/MGR stream : <- CONN tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768
2021/09/03 12:34:27.8132 debug   daemon/server-router/MGR stream : -> GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768
2021/09/03 12:34:27.8138 debug   daemon/server-router/MGR stream : <- CONN tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 11116
2021/09/03 12:34:27.8138 debug   daemon/server-router/MGR stream : -> GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 11116
2021/09/03 12:35:23.5708 debug   daemon/server-router/MGR stream : <- GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, code WRITE_CLOSED, len 0
2021/09/03 12:35:23.5709 debug   daemon/server-router/MGR stream : -- POOL tcp 10.42.0.10:54514 -> 127.0.0.1:8080, count now is 2
2021/09/03 12:35:23.6724 debug   daemon/server-router/MGR stream : -> GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, code READ_CLOSED, len 0
2021/09/03 12:35:38.2653 debug   daemon/server-router/TUN reader : <- TUN tcp 10.42.0.0:54281 -> 10.42.0.23:35835, sq 1434563858, an 991075845, wz 4096, len 0, flags=ACK
2021/09/03 12:35:38.2656 debug   daemon/server-router/TUN reader : -> MGR tcp 10.42.0.0:54281 -> 10.42.0.23:35835, code KEEP_ALIVE
2021/09/03 12:35:38.2657 debug   daemon/server-router/TUN writer : -> TUN tcp 10.42.0.23:35835 -> 10.42.0.0:54281, sq 991075845, an 1434563859, wz 4096, len 0, flags=ACK

# agent.log
time="2021-09-03 16:35:23.3199" level=debug msg="-> CONN tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.3452" level=debug msg="<- GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.3453" level=debug msg="-> CONN tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.4539" level=debug msg="<- GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.4539" level=debug msg="-> CONN tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.4838" level=debug msg="<- GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.4839" level=debug msg="-> CONN tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.6157" level=debug msg="<- GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.6159" level=debug msg="-> CONN tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.6169" level=debug msg="<- GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, len 32768" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.6170" level=debug msg="-> GRPC tcp 10.42.0.10:54514 -> 127.0.0.1:8080, code WRITE_CLOSED, len 0" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.6170" level=error msg="!! CONN tcp 10.42.0.10:54514 -> 127.0.0.1:8080, write: write tcp 10.42.0.23:9900->10.42.0.10:54514: write: broken pipe" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.6171" level=debug msg="-- POOL tcp 10.42.0.10:54514 -> 127.0.0.1:8080, count now is 2" THREAD=/forward lis=":9900"
time="2021-09-03 16:35:23.7147" level=debug msg="++ POOL tcp 10.42.0.10:54514 -> 127.0.0.1:8080, count now is 3" THREAD=/forward lis=":9900"

# traffic-manager.log
time="2021-09-03 16:35:23.0334" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.0741" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.1869" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.2213" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.3195" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.3448" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.4535" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.4832" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.6149" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.6165" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.6172" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to client, len 0" THREAD="/httpd/conn=10.42.0.22:8081" session_id=7c5e64dd-0ff7-469c-9c5d-23c1f7e9e548
time="2021-09-03 16:35:23.7136" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.7476" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.9024" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
time="2021-09-03 16:35:23.9285" level=debug msg=">> FRWD tcp 10.42.0.10:54514 -> 127.0.0.1:8080 to agent, len 32768" THREAD="/httpd/conn=127.0.0.1:8081" session_id=bc7cfcfb-9d91-4748-a030-7141be24cf9c
```

Notice how the daemon sends the last packet (with size `11116`) at ~`12:34:27`, but the traffic-manager is still forwarding packets of size `32768` at `13:35:23`, almost a full minute later.
The daemon has sent packets too fast and filled the stream with messages that the traffic manager has yet to forward to the agent.
In this context, at `16:35:23.6170`, a write to the pod fails with `broken pipe` because the pod has decided to abandon the request.
The agent emits a `WRITE_CLOSED` to inform the daemon to send no more data, and you can actually see it be forwarded by the traffic-manager at `16:35:23.6172`, and be received by the daemon at `12:35:23.5708` (**NOTE** I don't think the packet is travelling backwards through time, I suspect there may be slight clock skew between my machine and the cluster)
But because the daemon has already sent everything it was going to send, the `WRITE_CLOSED` doesn't really matter; the traffic-manager will continue reading from the [ClientTunnel and dispatching to the agent](https://github.com/telepresenceio/telepresence/blob/release/v2/cmd/traffic/cmd/manager/internal/state/state.go#L628-L661).

## What do?

I think there's multiple fixes that need to go in here:

1. Some safety on the agent side to prevent it from hanging. Something as simple as an early return from `HandleMessage` would prevent the problem where a `curl` fails even long after `wrk` is done running.
1. Some safety on the traffic-manager side to handle the pod hanging up before everything has been forwarded. If the agent is going to discard a message we give it because it has nowhere to send it to, then we shouldn't give it that message in the first place.
1. Maybe some backpressure tweaks are in order to prevent the daemon from dumping that much data into the tunnel if the manager is not going to be able to deal with it.
1. Some performance improvements along the whole chain? It's hard to say how exactly, but it'd be good if we had a better way of dealing with these large responses so that the congestion doesn't happen in the first place. Maybe we could implement some message batching or something.

**Comment by thallgren** [comment]
@josecv great investigation and conclusions. So very helpful. Thanks!

I think #2056 will address all the items in your "What do" list and I'm able to successfully run the `kubectl exec -ti ${WRK_POD_NAME} -- wrk -c 2 -t 2 -d 2m http://echo-easy.default/file` command, albeit with an added `--timeout 30s` since I got request timeouts without that (which is normal, I'm on a slow network).

**Comment by thallgren** [comment]
There are still issues with large file transfers. See #1953 and #2092.

**Comment by khussey** [comment]
This has been fixed in the 2.4.5 release, which is now available.
